marge/job.py

```diff
@@ -151,6 +151,37 @@ def rebase_and_accept(self, approvals):
             except gitlab.Unauthorized:
                 log.warning('Unauthorized!')
                 raise CannotMerge('My user cannot accept merge requests!')
+            except gitlab.NotFound as e:
+                log.warning('Not Found!: %s', e)
+                merge_request.refetch_info()
+                if merge_request.state == 'merged':
+                    # someone must have hit "merge when build succeeds" and we lost the race,
+                    # the branch is gone and we got a 404. Anyway, our job here is done.
```

Comment:
By logging you mean "leaving a message on the MR"? If we are executing this line is because marge's commits were already pushed and these contained the `Reviewed-by` tags and it is extremely likely that these commits are the ones merged. For this not to happen, as far as I can tell, the following events should take place:

  1. Marge has pushed the commits and is waiting for CI to pass
  1. Someone pushes with `--force` new commits, say, without the tags
  1. GitLab is configured not to stop the ongoing pipelines when a new version of the branch is pushed (otherwise Marge would notice the pipeline was cancelled)
  1. The CI pipeline for the new version of the branch finishes faster than Marge's

I'd optimise for the common case of someone pressing "Merge when build succeeds" since, missing `Reviewed-by` tags should be caught by a git-hook anyway!